rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalLanguageServices.java

```diff
@@ -235,6 +245,115 @@ public InterruptibleFuture<ITuple> getRename(ITree module, Position cursor, Set<
+                String newProject = StringUtils.removeStart(newWsFolder.toString(), commonProjPrefix);
+
+                throw new RuntimeException(String.format("Cannot automatically change uses of %s, since moving files between projects (from %s to %s) is not supported", currentLoc, currentProject, newProject));
+            }
+
+            PathConfig pcfg = getPathConfig.apply(currentWsFolder);
+            try {
+                IString currentName = VF.string(pcfg.getModuleName(currentLoc));
+                IString newName = VF.string(pcfg.getModuleName(newLoc));
+
+                writer.insert(VF.tuple(currentName, newName, addResources(pcfg)));
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        return writer.done();
+    }
+
+    private static String readFile(ISourceLocation loc, String charset) throws IOException {
+        URIResolverRegistry reg = URIResolverRegistry.getInstance();
+        try (Reader reader = reg.getCharacterReader(loc, charset)) {
+            StringBuilder res = new StringBuilder();
+            char[] chunk = new char[8192]; // from Prelude.java
+            int read;
+            while ((read = reader.read(chunk, 0, chunk.length)) != -1) {
+                res.append(chunk, 0, read);
+            }
+            return res.toString();
+        }
+    }
+
+    private CompletableFuture<ITree> getEditorTreeOrParse(ISourceLocation loc, Map<ISourceLocation, TextDocumentState> documents) throws IOException {
+        if(documents.containsKey(loc)) {
+            var file = documents.get(loc);
+            return file.getCurrentTreeAsync()
+                .thenApply(Versioned::get)
+                .handle((t, r) -> (t == null ? file.getMostRecentTree().get() : t))
+                ;
+        }
+
+        var input = readFile(loc, "UTF-8");
+        return parseSourceFile(loc, input);
+    }
+
+    public InterruptibleFuture<ITuple> getModuleRenames(List<FileRename> fileRenames, Set<ISourceLocation> workspaceFolders, Function<ISourceLocation, PathConfig> getPathConfig, Map<ISourceLocation, TextDocumentState> documents) {
+        var emptyResult = VF.tuple(VF.list(), VF.map());
+        if (fileRenames.isEmpty()) {
+            return InterruptibleFuture.completedFuture(emptyResult);
+        }
+
+        final ISet qualifiedNameChanges = qualfiedNameChangesFromRenames(fileRenames, workspaceFolders, getPathConfig);
+
+        return runEvaluator("Rascal module rename", semanticEvaluator, eval -> {
+            IFunction rascalGetPathConfig = eval.getFunctionValueFactory().function(getPathConfigType, (t, u) -> addResources(getPathConfig.apply((ISourceLocation) t[0])));
+            final var treeType = tf.abstractDataType(store, "Tree");
+            IFunction rascalGetModuleTree = eval.getFunctionValueFactory().function(tf.functionType(treeType, tf.tupleType(tf.sourceLocationType()), tf.tupleEmpty()), (args, kwArgs) -> {
```

Comment:
I get why this is here, so that we can look at unsaved state. But since the fix for #355 is on it's way, isn't this a bit of overkill? (especially since the rest of the rename code also doesn't work with unsaved changes I think?)

Reply:
Alright. Removed this.